ntimes.1
.TH "NTIMES" "1" " General Commands Manual" "ntimes" "JUNE 2016" 
.nh
.ad l


.SH NAME
.PP
ntimes \- \fBtime(1)\fP with average time, flaky rate, ..


.SH SYNOPSIS
.PP
\fBntimes\fP [OPTIONS] COMMAND [arg...]


.SH DESCRIPTION
.PP
ntimes is an improved version of \fBtime(1)\fP\&.

.PP
You can execute an command`N times, and measure the average/max/min time taken for the execution.
You can also measure the "flaky" rate (i.e. failure rate).


.SH OPTIONS
.PP
\fB\-\-help\fP
  Print usage statement

.PP
\fB\-f\fP, \fB\-\-format\fP=\fI""\fP
  Format string (in golang text/template, e.g. "{{json .}}")

.PP
\fB\-n\fP, \fB\-\-repeat\-n\-timesg\fP=\fI1\fP
  Number of times

.PP
\fB\-\-storage\fP
  Path to stdout,stderr storage

.PP
\fB\-\-version\fP
  Print version to stdout and exit

.PP
\fB\-\-warm\-up\fP
  Skip first n iterations for stat


.SH Examples
.PP
Example usage:

.PP
.RS

.nf
$ ntimes \-n 10 bash \-c 'sleep=$((RANDOM%5)); fail=$((RANDOM%2)); echo "id=$NTIMES\_ID, sleep=$sleep, fail=$fail"; sleep $sleep; exit $fail'
id=0, sleep=4, fail=0
id=1, sleep=0, fail=0
id=2, sleep=3, fail=1
id=3, sleep=4, fail=0
id=4, sleep=3, fail=1
id=5, sleep=0, fail=1
id=6, sleep=3, fail=1
id=7, sleep=0, fail=0
id=8, sleep=3, fail=1
id=9, sleep=3, fail=1

real average: 2.303595053s, max: 4.004607594s, min: 3.455164ms, std dev: 1.636182569s
real 99 percentile: 4.004607594s, 95 percentile: 4.004607594s, 50 percentile: 3.002190904s
user average: 0, max: 0, min: 0, std dev: 0
sys  average: 0, max: 0, min: 0, std dev: 0
flaky: 60%

.fi
.RE

.PP
You can specify the report format using Go's \fBtext/template\fP syntax.
Additionally to the standard functions provided by \fBtext/template\fP, the \fBjson\fP function is available.
Note that a \fBtime.Duration\fP value is expressed in nanoseconds.

.PP
.RS

.nf
$ ntimes \-\-format "{{json .}}" \-n 10 dd if=/dev/urandom of=/dev/null bs=512 count=1000
1000+0 records in
1000+0 records out
512000 bytes (512 kB, 500 KiB) copied, 0.0448978 s, 11.4 MB/s
...
{"real":{"average":44155207,"max":68222928,"min":38143407,"stddev":9421337,"percentiles":{"50":39855284,"95":68222928,"99":68222928}},"user":{"average":0,"max":0,"min":0,"stddev":0},"system":{"average":36000000,"max":36000000,"min":36000000,"stddev":0},"flaky":0}

.fi
.RE

.PP
Practical example for debugging flaky tests with Namazu ( \fBnmz(1)\fP, 
\[la]https://github.com/osrg/namazu):\[ra]

.PP
.RS

.nf
$ cd some\_maven\_project
$ sudo ntimes \-n 10 \-\-storage /tmp/logs nmz inspectors \-cmd "mvn test"
...
Flaky: 10%

$ find /tmp/logs \-name result.json | xargs jq .successful
true
true
true
false
true
true
true
true
true
true

.fi
.RE


.SH SEE ALSO
.PP
\fBtime(1)\fP, \fBnmz(1)\fP


.SH AUTHOR
.PP
June 2016, writen by Akihiro Suda

\[la]https://github.com/AkihiroSuda/ntimes\[ra]


ntimes.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"time"
)

type ntimes struct {
	N       int
	Cmd     *exec.Cmd
	Storage string
	WarmUp  int
	debug   bool
	results []Result
}

func (nt *ntimes) Run() error {
	if nt.Storage != "" {
		// MkdirAll does nothing if storage exists
		if err := os.MkdirAll(nt.Storage, 0755); err != nil {
			return err
		}
	}
	// we don't support parallelizing at the moment
	for id := 0; id < nt.N; id++ {
		if err := nt.prepareStorageDir(id); err != nil {
			return err
		}
		r, err := nt.run(id)
		if err != nil {
			// err is a really critical error.
			// if cmd failed, result.status will be
			// set to some error value, but err is
			// not returned.
			return fmt.Errorf("id %d: %+v", id, err)
		}
		if err = nt.storeResult(id, r); err != nil {
			return err
		}
		if id >= nt.WarmUp {
			nt.results = append(nt.results, r)
		} else {
			if nt.debug {
				log.Printf("id %d warm-up", id)
			}
		}
	}
	return nil
}

func fileExists(path string) (bool, error) {
	_, err := os.Stat(path)
	if err == nil {
		return true, nil
	}
	if os.IsNotExist(err) {
		return false, nil
	}
	return true, err
}

func (nt *ntimes) storageDir(id int) string {
	if nt.Storage != "" {
		return filepath.Join(nt.Storage, strconv.Itoa(id))
	}
	return ""
}

func (nt *ntimes) prepareStorageDir(id int) error {
	dir := nt.storageDir(id)
	if dir != "" {
		exists, err := fileExists(dir)
		if err != nil {
			return err
		}
		if exists {
			return os.ErrExist
		}
		if err := os.Mkdir(dir, 0755); err != nil {
			return err
		}
	}
	return nil
}

func (nt *ntimes) prepareStdio(id int) (io.Reader, io.Writer, io.Writer, error) {
	dir := nt.storageDir(id)
	if dir != "" {
		stdout, err := os.Create(filepath.Join(dir, "stdout"))
		if err != nil {
			return nil, nil, nil, err
		}
		stderr, err := os.Create(filepath.Join(dir, "stderr"))
		if err != nil {
			return nil, nil, nil, err
		}
		return nt.Cmd.Stdin, stdout, stderr, nil
	}
	return nt.Cmd.Stdin, nt.Cmd.Stdout, nt.Cmd.Stderr, nil
}

func (nt *ntimes) prepareCmd(id int) (*exec.Cmd, error) {
	if nt.Cmd.ExtraFiles != nil {
		err := fmt.Errorf("ExtraFiles is not supported: %v", nt.Cmd)
		return nil, err
	}
	stdin, stdout, stderr, err := nt.prepareStdio(id)
	if err != nil {
		return nil, err
	}
	env := append(nt.Cmd.Env, fmt.Sprintf("NTIMES_ID=%d", id))
	cmd := &exec.Cmd{
		Path:   nt.Cmd.Path,
		Args:   nt.Cmd.Args,
		Env:    env,
		Dir:    nt.Cmd.Dir,
		Stdin:  stdin,
		Stdout: stdout,
		Stderr: stderr,
	}
	return cmd, nil
}

func (nt *ntimes) run(id int) (Result, error) {
	r := Result{ID: id}
	cmd, err := nt.prepareCmd(id)
	if err != nil {
		return r, err
	}
	begin := time.Now()
	err = cmd.Run()
	end := time.Now()

	ps := cmd.ProcessState
	if ps == nil {
		if err != nil {
			return r, err
		}
		return r, fmt.Errorf("ProcessState is nil for %#v", cmd)
	}
	r.Successful = ps.Success()
	r.Real, r.User, r.System = end.Sub(begin), ps.UserTime(), ps.SystemTime()

	// maybe ps is useful. so we check err here.
	if err != nil {
		if _, ok := err.(*exec.ExitError); !ok {
			return r, err
		}
	}
	r.status, r.rusage = ps.Sys(), ps.SysUsage()
	if nt.Storage != "" {
		if err = cmd.Stdout.(io.WriteCloser).Close(); err != nil {
			return r, err
		}
		if err = cmd.Stderr.(io.WriteCloser).Close(); err != nil {
			return r, err
		}
	}
	return r, nil
}

func (nt *ntimes) storeResult(id int, result Result) error {
	if id != result.ID {
		return fmt.Errorf("id %d mismatch: %+v", id, result)
	}
	dir := nt.storageDir(id)
	if dir == "" {
		// NOP, not an error
		return nil
	}
	file, err := os.Create(filepath.Join(dir, "result.json"))
	if err != nil {
		return err
	}
	enc := json.NewEncoder(file)
	if err = enc.Encode(result); err != nil {
		return err
	}
	return file.Close()
}

// Stat returns a Stat.
// This may change the order of nt.results.
func (nt *ntimes) Stat() (*Stat, error) {
	stat := &Stat{}
	sels := []selector{selReal, selUser, selSystem}
	stat.Real, stat.User, stat.System =
		&TimeStat{}, &TimeStat{}, &TimeStat{}
	updateStatPhase0(stat, nt.results, sels)
	updateStatPhase1(stat, nt.results, sels)

	// TODO: support percentiles for user and sys
	sort.Sort(resultsByReal(nt.results))
	stat.Real.Percentiles = make(map[string]time.Duration, 0)
	for _, i := range []int{99, 95, 50} {
		p, err := percentile(nt.results, i, selReal)
		if err != nil {
			return stat, err
		}
		stat.Real.Percentiles[strconv.Itoa(i)] = p
	}
	return stat, nil
}
